contracts/sui/xcall/sources/main.move

```diff
@@ -0,0 +1,547 @@
+            xcall_state::set_call_reply(self,msg);
+            fee
+        } else{
+            let sendSn = if (need_response) {sequence_no} else {0};
+            let cs_message=cs_message::from_message_request(cs_request);
+            
+       
+            connection_send_message(self,
+            fee,
+            envelope::sources(&envelope),
+            network_address::net_id(&to),
+            cs_message,
+            sendSn,
+            false,
+            ctx)
+        };
+
+        event::emit(CallMessageSent{from:network_address::addr(&from),to:network_address::to_string(&to),sn:sequence_no});  
+        fee    
+    }
+
+    fun connection_send_message(self:&mut Storage,mut fee:Coin<SUI>,sources:vector<String>, net_to:String, cs_message:CSMessage,sn:u128,is_response:bool,ctx: &mut TxContext):Coin<SUI>{
+        let mut protocols=sources;
+        if(vector::is_empty(&sources)){
+            let connection= xcall_state::get_connection(self,net_to);
+            vector::push_back(&mut protocols,connection);
+           
+        };
+        
+            let cs_message_bytes=cs_message::encode(&cs_message);
+            let mut i=0;
+            while(i < vector::length(&protocols)){
+                let protocol=*vector::borrow(&protocols,i);
+
+                let required_fee = get_connection_fee(self, protocol, net_to, is_response);
+
+                let paid= fee.split(required_fee,ctx);
+                connections::send_message(
+                    xcall_state::get_connection_states_mut(self),
+                    protocol,
+                    paid,
+                    net_to,
+                    sn,
+                    cs_message_bytes,
+                    is_response,
+                    ctx);
+
+                i=i+1;
+            };
+
+            fee
+    }
+
+    fun get_next_sequence(self:&mut Storage):u128 {
+        let sn=xcall_state::get_next_sequence(self);
+        sn
+    }
+
+
+    fun get_next_req_id(self:&mut Storage):u128 {
+        let req_id=xcall_state::get_next_request_id(self);
+        req_id
+    }
+
+    entry fun set_protocol_fee(self:&mut Storage,admin:&AdminCap,fee:u64){
+        self.enforce_version(CURRENT_VERSION);
+        xcall_state::set_protocol_fee(self,fee);
+    }
+
+    entry fun set_protocol_fee_handler(self:&mut Storage,admin:&AdminCap,fee_handler:address){
+        self.enforce_version(CURRENT_VERSION);
+        xcall_state::set_protocol_fee_handler(self,fee_handler);
+    }
+
+    public fun send_call(self:&mut Storage,fee: Coin<SUI>,idCap:&IDCap,to:String,envelope_bytes:vector<u8>,ctx: &mut TxContext){
+        let envelope=envelope::decode(&envelope_bytes);
+        let to = network_address::from_string(to);
+        let from= network_address::create(xcall_state::get_net_id(self),utils::id_to_hex_string(&object::id(idCap)));
+
+        let remaining = send_call_inner(self,fee,from,to,envelope,ctx);
+        transfer::public_transfer(remaining, ctx.sender());
+    }
```

Comment:
What happens when a user where to call this directly? 
I am guessing the can't specify a idCap?

Also in XCall users as supposed to directly send messages where the from would be their network address. From the looks of this, that wont be possible

Reply:
ya this can be called by anyone but they would have to register themselves as dapp which will give them a unique id and that will be used to get a `from` address. i think we can make direct calls work as well using sender address but they wont be able to execute call without some code or package deployed to handle the ticket so makes no sense.